In [1]:
# Behzad Ghabaei
# CS 82A - Data Science
# module 4 lab 2 
# TextBook Problem 3.1
# module4_lab2.ipynb
# Instructor Seno
# 9/23/2026

In [2]:
import pandas as pd
df = pd.read_csv("allbacks.csv")

#### We will be observing 15 books, a small sample size, giving the **volume, area, weight, and cover.** There are 15 instances of the 4 attributes.
#### -Volume: book volumes in cubic centimeters.
#### -Area: total area of book in square centimenters.
#### -Weight: book's weight in grams.
#### -Cover: Hb for hardback, and Pb for paperback.
#### Which type of book do we want to procure in the future?
#### 1. cover median: For the books data: find the median cover type (hint: a median works on ordered categories too). For the median of a category,  sort the values and take the middle one, exactly as with numbers.  The by-hand verification matches Python.
#### 2. weight mean: the mean weight of all books (overall),  and the mean weight within each cover type (hardcover vs. paperback). Compare the group means and say what the difference suggests.
#### 3. volume variance
#### Which type of cover is best?
#### All three statistics each gets one plain-English sentence, verify at least one by hand.

## Step 1: First lets run these 4 commands

In [3]:
print(df.shape)
print("----")
print(df.info())
print("----")
print(df.describe())
print("----")
print(df.head(15))
print("----")

(15, 5)
----
<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   rownames  15 non-null     int64
 1   volume    15 non-null     int64
 2   area      15 non-null     int64
 3   weight    15 non-null     int64
 4   cover     15 non-null     str  
dtypes: int64(4), str(1)
memory usage: 732.0 bytes
None
----
        rownames       volume        area       weight
count  15.000000    15.000000   15.000000    15.000000
mean    8.000000   845.266667  182.800000   706.666667
std     4.472136   339.879909  203.586837   268.837887
min     1.000000   239.000000    0.000000   250.000000
25%     4.500000   618.000000    0.000000   512.500000
50%     8.000000   929.000000    0.000000   725.000000
75%    11.500000  1025.000000  376.500000   950.000000
max    15.000000  1492.000000  468.000000  1075.000000
----
    rownames  volume  area  weight cover
0          1     885   382     800 

## Step 2: lets solve the median of "cover" by hand.

#### 1. solving the str cover by hand we find that the "cover" median has:
#### 15 books in this list, when sorted the one in the middle is the paper back.
#### 1h 2h 3h 4h 5h 6h 7h **8p** 9p 10p 11p 12p 13p 14p 15p
#### 1p 2p 3p 4p 5p 6p 7p **8p** 9h 10h 11h 12h 13h 14h 15h
#### so the median is **paperback** the hand verification matches python.

In [4]:
len(df["cover"])

15

In [5]:
len(df[df["cover"] == "hb"])

7

In [6]:
len(df[df["cover"] == "pb"])

8

#### Solving by python we can get the median of 7, 8 and this returns 7.5 so paperback was correct.  We can use lecture 2.2 Collections and Loops to sort the list with sorted() and get the index in the center.  Although strings are sorted in alphabetical order, the idea is to choose the central index.

In [7]:
import statistics
my_list = [8,7]
# the median
result = statistics.median(my_list)
print(result) 

7.5


In [8]:
cover_list = ["p","p","p","p","p","p","p","p","h","h","h","h","h","h","h"]
new_list = sorted(cover_list)
print("index: ",15//2)
print(new_list[7]) # paperback

index:  7
p


#### We verified that the median of the str data is the paperback which was the index in the center.  So perhaps most books at the library are paper back.  More than half are most likely magazines, short novels, and short stories, and less than half are reference books and text books.

## Step 3: We are looking for the weight mean
#### The overall can be found with the describe() function, and using the groupby() function we can get the individual mean for each category inside the cover column, hard back and paper back.

#### We can't change rownames to a string, because the groupby function will not run for two strings one being "cover" and the other being "rownames", and we should not drop rownames from the mean and median collection. although it will not appear in the tables the output may be affected.

#### We could use groupby() function that can split, calculate and combine.  In this example, each row from "cover" (hb, pb) is split apart by str label, the median of the numerical values are separately calculated in that column classification, then data is joined together into a table. The weight median is possible to find this way also.

#### Here is the weight mean of each category hb and pb

In [9]:
df.groupby("cover").mean()

,rownames,volume,area,weight
cover,,,,
hb,4.0,833.571429,391.714286,796.428571
pb,11.5,855.500000,0.000000,628.125000


#### hard back weight mean is 796.42
#### paper back weight mean is 628.12

In [10]:
print("mean weight: \n", df["weight"].mean())

mean weight: 
 706.6666666666666


#### The overall weight mean is 706.66, the describe also verifies this mean. It also points out the median and min area of paperbacks are 0.

In [11]:
df.describe()

,rownames,volume,area,weight
count,15.000000,15.000000,15.000000,15.000000
mean,8.000000,845.266667,182.800000,706.666667
std,4.472136,339.879909,203.586837,268.837887
min,1.000000,239.000000,0.000000,250.000000
25%,4.500000,618.000000,0.000000,512.500000
50%,8.000000,929.000000,0.000000,725.000000
75%,11.500000,1025.000000,376.500000,950.000000
max,15.000000,1492.000000,468.000000,1075.000000


#### The decsribe function also confirms the overall weight mean.  Knowing the mean overall versus category, tells us that the paperback is lighter and hardback is heavier.  Paper back weight mean is 628.12 g, hard back weight mean is 796.42 g, and the overall weight mean is 706.66 g.  A difference of about 168 grams on each book can occur if the books become hard back from paper back. This is considerably heavier on the book shelves, and requires much more careful pulling out and shelving.  Strollers will be essential for re-shelving the returning books or newly incoming books, back to their shelves, while checking out or purchasing large numbers of books can become a hefty experience.  Paper back books are fragile but lighter. They may however need hard backs to last longer.  We also noticed that all the area data is missing from paperbacks.  They are all listed as 0. Area is width * height. If this was a book store. The data says that heavier books could be more interesting for customers than lighter books, because it offers something that they could share with large groups of people.  The difference in weight is not that great, and this indicates that offering customers a hard book option may be a good idea, since more than half are paperback.

In [12]:
print("The mean difference in weight (grams)", 796.42 - 628.17)

The mean difference in weight (grams) 168.25


## Step 4: The volume variance

#### The variance is used to measure how spread the data points are. First we pick a center distribution like the mean and measure distance from each data point. Variance as a measure, gives us the measure of spread in units squared.  Taking the square root of variance returns the standard deviation.  The two types of variance are variance of a population, and variance of a sample. The variance of volume is very large at 115518.35

In [13]:
df["volume"].var()

np.float64(115518.3523809524)

In [14]:
df["volume"].std()

np.float64(339.8799087633048)

#### Due to the standard deviation looking high the median is a good way of deciding on which cover to choose.  The variance says that the fluctuation of volume occurs with paper back and hard back, which could mean there is a large variety of materials that are very different from each other.  High variance also indicates problems with stacking books.  Some books are very thick and some are very thin. The volume is measured in cubic centimeters and the sizes being different, will affect the organizational appearance on the shelves.  Some will be tall and thin, and others will be short and thick.  Choosing a hard back plan for the books at the library could save the problem of high variance.  But the weight will be considerably heavier, of about 168 grams more.  Choosing paperback is lighter and most books are this anyway.  A good question is, if there is a relationship between hard back and paper back, can we use it to find the 0 areas in paper back.  In conclusion leaving the backs as paper back saves the library and book store trouble and keeps the books lighter.

#### Behzad Ghabaei - End Of Lab